### PR TITLE
ESPnet2 recipe for aidatatang

### PR DESCRIPTION
**Pull request by nzhoward**
Currently, the performance of this espnet2 recipe does not reach that of the espnet1 recipe.
We can discuss ways to tune the ESPnet2 recipe.

- Differences
  - The current input feature in this recipe is the **same** (fbank + pitch) extracted by Kaldi. 
    - However, the purpose of espnet2 is to exclude the Kaldi dependency as much as possible, and we need to investigate the default espnet2 feature extraction later.
    - The default espnet2 feature extraction setup is very different (different STFT configurations and no pitch feature) from the espnet1 one, and we may need to adjust it to espnet1.
    - Also, some hyper-parameters (batch size, cut of the long utterances etc.) are affected by the feature difference.
  - The optimizer hyper-parameters are different (so-called [noamlr](https://github.com/espnet/espnet/blob/536d163870e29ea52bbaa4cc0b03cfc407df0c0b/espnet2/schedulers/noam_lr.py) and [warmuplr](https://github.com/espnet/espnet/blob/536d163870e29ea52bbaa4cc0b03cfc407df0c0b/espnet2/schedulers/warmup_lr.py)

Suggestions:
- [ ] switch back to the espnet1 equivalent optimizer setup (noamlr) to make sure that we can reproduce the same result
- [ ] default warmuplr optimizer is not well-tuned. In this recipe, we observed the performance drop in the early stage of the epochs, which happens when the learning rate is too large in a certain iteration
    - [ ] @ShigekiKarita suggested, probably it's better to increase the batchsize instead of reducing the learning rate. But if the batch size reaches the maximum GPU memory, we could increase `accmu-grad` from 2 to 3.
    - [ ] reduce the learning rate
    - [ ] reduce the warmup stage (currently 25k)

### CODE REVIEW

**Comment by sw005320** [comment]
Can you make a separate PR for espnet1 and espnet2?
We can merge an espnet1 PR quickly, and continue the development in an espnet2 PR.

**Comment by nzhoward** [comment]
> Can you make a separate PR for espnet1 and espnet2?
> We can merge an espnet1 PR quickly, and continue the development in an espnet2 PR.

New PR for ESPnet1 is here
https://github.com/espnet/espnet/pull/2269

**Comment by sw005320** [comment]
@nzhoward, I just summarized our discussions.
Please follow our action items and perform experiments.
Please also paste the learning curves often so that we can have more concrete ideas of how we tune the model.

**Comment by nzhoward** [comment]
- Switch to ESPnet1 noamlr optimizer
- Performance is slightly worse than warmuplr
- Trained for 50 epochs with below configs:
```
# network architecture
# encoder related
encoder: transformer
encoder_conf:
    output_size: 256    # dimension of attention
    attention_heads: 4
    linear_units: 2048  # the number of units of position-wise feed forward
    num_blocks: 12      # the number of encoder blocks
    dropout_rate: 0.1
    positional_dropout_rate: 0.1
    attention_dropout_rate: 0.0
    input_layer: conv2d # encoder architecture type
    normalize_before: true

# decoder related
decoder: transformer
decoder_conf:
    attention_heads: 4
    linear_units: 2048
    num_blocks: 6
    dropout_rate: 0.1
    positional_dropout_rate: 0.1
    self_attention_dropout_rate: 0.0
    src_attention_dropout_rate: 0.0

# hybrid CTC/attention
model_conf:
    ctc_weight: 0.3
    lsm_weight: 0.1     # label smoothing option
    length_normalized_loss: false

# minibatch related
batch_type: folded
batch_size: 32

# optimization related
accum_grad: 3
grad_clip: 5
patience: 5
max_epoch: 50
val_scheduler_criterion:
    - valid
    - acc
best_model_criterion:
-   - valid
    - acc
    - max
keep_nbest_models: 10

optim: adam
optim_conf:
    lr: 0.1
scheduler: noamlr
scheduler_conf:
    model_size: 256
    warmup_steps: 25000
```

image
image

### Results

|dataset|Snt|Wrd|Corr|Sub|Del|Ins|Err|S.Err|
|---|---|---|---|---|---|---|---|---|
|dev|24216|234524|90.7|8.1|1.2|0.2|9.5|41.8|
|test|48144|468933|89.9|8.9|1.2|0.2|10.3|44.3|

**Comment by kamo-naoyuki** [comment]
Hi, noamlr is exactly same scheduler with warmup except for the scaling. The result must be same.

**Comment by sw005320** [comment]
@nzhoward, can you revisit this PR?
We actually have several optimal configs for espnets now (e.g., https://github.com/espnet/espnet/blob/master/egs2/aishell/asr1/README.md#conformer--specaug--speed-perturbation-featsraw-n_fft512-hop_length128, https://github.com/espnet/espnet/blob/master/egs2/librispeech/asr1/README.md#tuning-warmup_steps), which can get comparable/superior performance to espnet1 ones.

**Comment by sw005320** [comment]
Sorry, I was confused about this PR.
You can just ignore my previous comment.
